### PR TITLE
Huang (Princeton) + optuna

### PR DESCRIPTION
**Pull request by viktorvaladi**
new attack with option to use optuna for hyperparameter tuning

### CODE REVIEW

File: leakpro/fl_utils/gia_module_to_functional.py [hunk 1]
```diff
@@ -43,7 +42,7 @@ def __init__(self: Self, net: nn.Module) -> None:
         """Init with network."""
         gpu_or_cpu = device("cuda" if cuda.is_available() else "cpu")
         super().__init__()
-        self.net = deepcopy(net)
+        self.net = net
```
Comment (johanos1) [inline leakpro/fl_utils/gia_module_to_functional.py:45]: Blir det här verkligen rätt? net kommer skrivas över via ändringar i gia_module_to_functional nu

File: leakpro/fl_utils/img_utils.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 """Diverse util functions."""
 import torch.nn.functional as f
-from torch import Tensor, abs, cuda, mean, nn, no_grad
+from ignite.metrics import SSIM
```
Comment (johanos1) [inline leakpro/fl_utils/img_utils.py:3]: behöver läggas till i requirements?

File: leakpro/fl_utils/img_utils.py [hunk 3]
```diff
@@ -46,10 +52,44 @@ def dataloaders_psnr(original_dataloader: DataLoader, recreated_dataloader: Data
...
+
+    with no_grad():
+        # Zip through both dataloaders
+        for (orig_batch, rec_batch) in zip(original_dataloader, recreated_dataloader):
+            orig_images = orig_batch[0].to(device)  # Original images
+            rec_images = rec_batch[0].to(device)    # Recreated images
+
+            # Update SSIM metric
+            ssim_metric.update((rec_images, orig_images))
+            total_images += orig_images.size(0)
```
Comment (johanos1) [inline leakpro/fl_utils/img_utils.py:83]: this is not used anywhere.

File: leakpro/fl_utils/img_utils.py [hunk 4]
```diff
@@ -46,10 +52,44 @@ def dataloaders_psnr(original_dataloader: DataLoader, recreated_dataloader: Data
...
+
+    # Compute average SSIM
+    return ssim_metric.compute()
+
 class MedianPool2d(nn.Module):
```
Comment (johanos1) [inline leakpro/fl_utils/img_utils.py:88]: Varför behövs denna i img_util?

File: examples/gia/huang_optuna/main.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
+"""Hyperparameter tuning with optuna on evaluating."""
+
+from cifar import get_cifar10_loader
+
+from leakpro.fl_utils.gia_train import train
+from leakpro.run import huang_optuna
```
Comment (johanos1) [inline examples/gia/huang_optuna/main.py:6]: Inget fan av att populera Leakpro med en massa olika run-funktioner såhär.

File: examples/gia/huang_cifar10_16_images/model.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
+from torch import nn
```
Comment (johanos1) [inline examples/gia/huang_cifar10_16_images/model.py:1]: Om den här koden tagits någonstans ifrån bör det finnas en referens här.

File: leakpro/fl_utils/model_utils.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        self.mean = mean
+        self.var = var
+        self.r_feature = r_feature
+
+    def close(self: Self) -> None:
+        """Remove the hook."""
+        self.hook.remove()
+
+
+def seed_everything(seed: int) -> None:
```
Comment (johanos1) [inline leakpro/fl_utils/model_utils.py:40]: detta skulle kunna läggas på ett mer generellt ställe.

File: leakpro/fl_utils/model_utils.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    Will compute mean and variance, and will use l2 as a loss.
+    """
+
+    def __init__(self: Self, module: torch.nn.modules.BatchNorm2d) -> None:
+        self.hook = module.register_forward_hook(self.hook_fn)
+
+    def hook_fn(self: Self, module: torch.nn.modules.BatchNorm2d, input: torch.Tensor, _: torch.Tensor) -> None:
+        """Hook to compute deepinversion's feature distribution regularization."""
+        nch = input[0].shape[1]
+        mean = input[0].mean([0, 2, 3])
```
Comment (johanos1) [inline leakpro/fl_utils/model_utils.py:23]: kommentera varför du väljer dessa index.

File: leakpro/attacks/gia_attacks/huang.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+    at_iterations: int = 10000
+    # MetaOptimizer, see MetaSGD for implementation
+    optimizer: object = field(default_factory=lambda: MetaSGD())
+    # Client loss function
+    criterion: object = field(default_factory=lambda: CrossEntropyLoss(reduction="mean"))
+    # Number of epochs for the client attack
+    epochs: int = 1
+    # if to use median pool 2d on images, can improve attack on high higher resolution (100+)
+    median_pooling: bool = False
+    # bn regulizer
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/huang.py:40]: spelling mistake.

File: leakpro/attacks/gia_attacks/huang.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+        self.loss_r_feature_layers = []
+
+        for module in self.model.modules():
+            if isinstance(module, torch.nn.BatchNorm2d):
+                self.loss_r_feature_layers.append(BNFeatureHook(module))
+        # calculate running bn statistics
+        _ = self.train_fn(self.model, self.client_loader, self.optimizer, self.criterion, self.epochs)
+
+        # Stop updating running statistics
+        for module in self.model.modules():
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/huang.py:109]: Is it enough to have momentum=0?

File: leakpro/attacks/gia_attacks/huang.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+        Returns:
+        -------
+            None
+
+        """
+        # add feature hook
+        self.loss_r_feature_layers = []
+
+        for module in self.model.modules():
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/huang.py:102]: comment

File: leakpro/attacks/gia_attacks/huang.py [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+            if isinstance(module, torch.nn.BatchNorm2d):
+                self.loss_r_feature_layers.append(BNFeatureHook(module))
+        # calculate running bn statistics
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/huang.py:105]: This is simulating the client, right? Wouldnt it make sense to do it only once in the set up of attacks and have all FL attacks use the same client model?

File: leakpro/attacks/gia_attacks/huang.py [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+        _ = self.train_fn(self.model, self.client_loader, self.optimizer, self.criterion, self.epochs)
+
+        # Stop updating running statistics
+        for module in self.model.modules():
+            if isinstance(module, torch.nn.BatchNorm2d):
+                module.momentum = 0
+
+        self.reconstruction, self.reconstruction_loader = get_at_images(self.client_loader)
+        self.reconstruction.requires_grad = True
+        client_gradient = self.train_fn(self.model, self.client_loader, self.optimizer, self.criterion, self.epochs)
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/huang.py:115]: Cant we use the gradient from line 106?

File: leakpro/attacks/gia_attacks/huang.py [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+        self.client_gradient = [p.detach() for p in client_gradient]
+
+
+    def run_attack(self:Self) -> Union[GIAResults, Generator[tuple[int, Tensor]]]:
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/huang.py:119]: This is almost identical to the Geiping attack. Would it make sense to bring it out and share it?

File: leakpro/attacks/gia_attacks/huang.py [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+            rec_loss += self.t_v_scale * total_variation(self.reconstruction)
+            rec_loss += self.bn_reg * loss_r_feature
+            rec_loss += self.l2_scale * l2_norm(self.reconstruction)
+            rec_loss.backward()
+            self.reconstruction.grad.sign_()
+            return rec_loss
+        return closure
+
+    def reconstruction_costs(self: Self, client_gradient: torch.Tensor, reconstruction_gradient: torch.Tensor) -> torch.Tensor:
+        """Computes the reconstruction costs between client gradients and the reconstruction gradient.
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/huang.py:198]: The description is identical to the Geiping code but the code is not. Why is that?

File: leakpro/attacks/gia_attacks/huang.py [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+
+        return GIAResults(self.client_loader, self.best_reconstruction,
+                          ssim_score=dataloaders_ssim_ignite(self.client_loader, self.reconstruction_loader),
+                          data_mean=self.data_mean, data_std=self.data_std)
+
+
+    def gradient_closure(self: Self, optimizer: torch.optim.Optimizer) -> Callable:
+        """Returns a closure function that performs a gradient descent step.
+
+        The closure function computes the gradients, calculates the reconstruction loss,
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/huang.py:162]: Update this comment, it is copied from the other attack. Explain the loss_r_feature.

File: leakpro/attacks/gia_attacks/huang.py [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+            # loss function which does training and compares distance from reconstruction training to the real training.
+            closure = self.gradient_closure(optimizer)
+
+            loss = optimizer.step(closure)
+            scheduler.step()
+            with torch.no_grad():
+                self.reconstruction.data = torch.max(
+                    torch.min(self.reconstruction, (1 - self.data_mean) / self.data_std), -self.data_mean / self.data_std
+                    )
+                if (i +1) % 500 == 0 and self.median_pooling:
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/huang.py:142]: what is going on here?

File: leakpro/attacks/gia_attacks/huang.py [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+                    self.reconstruction.data = MedianPool2d(kernel_size=3, stride=1, padding=1, same=False)(self.reconstruction)
+            if i % 250 == 0:
+                logger.info(f"Iteration {i}, loss {loss}")
+                yield i, dataloaders_ssim_ignite(self.client_loader, self.reconstruction_loader)
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/huang.py:146]: what is going on here? Please comment. Why is it a generator?

**Review by johanos1 — CHANGES_REQUESTED** [review]

File: leakpro/fl_utils/gia_module_to_functional.py (hunk 1 above)
Comment (viktorvaladi) [reply leakpro/fl_utils/gia_module_to_functional.py:45]: Yes blir rätt, vi är beorende på att det är samma net i huang attacken (för statistics), och i geiping så rör vi inte något i modellen ändå

File: leakpro/fl_utils/img_utils.py (hunk 4 above)
Comment (viktorvaladi) [reply leakpro/fl_utils/img_utils.py:88]: ssim?

File: leakpro/attacks/gia_attacks/huang.py (hunk 14 above)
Comment (viktorvaladi) [reply leakpro/attacks/gia_attacks/huang.py:119]: Yes I agree we should probably move it to abstractGIA for starters at least, but we should do it in a separate PR i think

File: leakpro/attacks/gia_attacks/huang.py (hunk 15 above)
Comment (viktorvaladi) [reply leakpro/attacks/gia_attacks/huang.py:198]: They do the same thing but I feel like the geiping one should not work, since they do + cost instead of minus cost, but for some reason it does work anyway. I think we should investigate it more and make a generic shared function for cosine similarity here

Comment (viktorvaladi) [reply leakpro/attacks/gia_attacks/huang.py:198]: ill look in to it

File: leakpro/attacks/gia_attacks/huang.py (hunk 18 above)
Comment (viktorvaladi) [reply leakpro/attacks/gia_attacks/huang.py:146]: Its for optuna to allow early stopping, ill add a comment. could probably do a little improvement here as well but i think also for another PR where we move this to abstractgia

Comment (johanos1) [reply leakpro/attacks/gia_attacks/huang.py:146]: Ok, add that as an issue.

File: leakpro/attacks/gia_attacks/huang.py (hunk 15 above)
Comment (johanos1) [reply leakpro/attacks/gia_attacks/huang.py:198]: Yes, I think that should be part of this PR too. We create too much technical debt if we just push everything forward.

File: leakpro/attacks/gia_attacks/huang.py (hunk 14 above)
Comment (johanos1) [reply leakpro/attacks/gia_attacks/huang.py:119]: then create an issue.

File: leakpro/fl_utils/gia_module_to_functional.py (hunk 1 above)
Comment (johanos1) [reply leakpro/fl_utils/gia_module_to_functional.py:45]: men här ändrar du ju self.net även för geiping?

File: leakpro/fl_utils/img_utils.py (hunk 4 above)
Comment (johanos1) [reply leakpro/fl_utils/img_utils.py:88]: men SSIM är baserad på medelvärden och varians, inte median? https://en.wikipedia.org/wiki/Structural_similarity_index_measure

File: leakpro/fl_utils/gia_module_to_functional.py (hunk 1 above)
Comment (viktorvaladi) [reply leakpro/fl_utils/gia_module_to_functional.py:45]: Inget ändras på self.net, modulerna byts ut mot functional och sen byts de tillbaka till orginal modulen. Eftersom vi inte uppdaterar orginal vikterna under träningen så behövs inte deepcopy. Uppdaterar alltid till en ny kopia av vikter så self.net förblir samma

File: leakpro/fl_utils/img_utils.py (hunk 4 above)
Comment (viktorvaladi) [reply leakpro/fl_utils/img_utils.py:88]: Ja det stämmer men hur relaterar det till koden?

File: leakpro/fl_utils/model_utils.py (hunk 7 above)
Comment (viktorvaladi) [reply leakpro/fl_utils/model_utils.py:40]: något förslag på ställe för den?

File: leakpro/attacks/gia_attacks/huang.py (hunk 10 above)
Comment (viktorvaladi) [reply leakpro/attacks/gia_attacks/huang.py:109]: yes! we only care about the running statistics not updating

File: leakpro/attacks/gia_attacks/huang.py (hunk 17 above)
Comment (viktorvaladi) [reply leakpro/attacks/gia_attacks/huang.py:142]: forcing values of pixels into a reasonable range

File: leakpro/attacks/gia_attacks/huang.py (hunk 12 above)
Comment (viktorvaladi) [reply leakpro/attacks/gia_attacks/huang.py:105]: yea but this differs between attacks so would not be worth the complexity id say. The clients train under different settings in inverting and huang

File: leakpro/fl_utils/img_utils.py (hunk 4 above)
Comment (johanos1) [reply leakpro/fl_utils/img_utils.py:88]: jag fattar inte. Du använder inte `MedianPool2d` i `dataloaders_ssim_ignite`? Enda gången den klassen används är var 500e iteration i både geiping och huang attackerna? Var 500e iteration gör du `self.reconstruction.data` till ett MedianPool2d objekt. Du anropar dock SSIM var 250e iteration vilket innebär att ssim räknas även ut när `self.reconstruction.data` inte är ett MedianPool2d objekt? Varför behövs MedianPool2d?

File: leakpro/fl_utils/model_utils.py (hunk 7 above)
Comment (johanos1) [reply leakpro/fl_utils/model_utils.py:40]: gör en python-fil i leakpro/utils

File: leakpro/attacks/gia_attacks/huang.py (hunk 17 above)
Comment (johanos1) [reply leakpro/attacks/gia_attacks/huang.py:142]: Hmm. You are effectively low-pass filtering the image by doing this. I mean, you simply replace a pixel value with the median of a 3x3 window with that pixel in the center.

File: leakpro/fl_utils/img_utils.py (hunk 4 above)
Comment (viktorvaladi) [reply leakpro/fl_utils/img_utils.py:88]: Jaha, trodde din kommentar var på ssim funktionen. De är inte relaterade till varandra. MedianPool2d är bra att köra när det är hög resolution på bilderna.

File: leakpro/attacks/gia_attacks/huang.py (hunk 17 above)
Comment (viktorvaladi) [reply leakpro/attacks/gia_attacks/huang.py:142]: your comments are being placed at strange places i think. 
image
this looks like its on the torch.max operation your commenting on.

The median pooling is needed for "inverting gradients" to work on higher resolution images such as celebA / imagenet. My guess is it helps to by making some guesses in the reconstruction which can help to make it better with the reconstruction step on next iterations

**Review by johanos1 — APPROVED** [review]
